opto/features/priority_search/sampler.py

```diff
@@ -0,0 +1,314 @@
+                info=infos[_index],
+                target=outputs[_index][0],  # target output
+                score=outputs[_index][1],  # score of the proposal
+                feedback=outputs[_index][2],  # feedback of the proposal
+            )
+            _index += 1  # increment the index
+            rollouts.append(rollout)
+        results.append(RolloutsGraph(rollouts))  # append the subgraph to the results
+    return results
+
+
+
+class Sampler:
+
+    """ A sampler that samples a batch of data from the loader and evaluates the agents on the sampled inputs.
+    """
+    def __init__(self, loader, guide, num_threads=1, sub_batch_size=None, forward=None, score_range=(-np.inf, np.inf)):
+        """ Initialize the sampler with a data loader and a guide.
+
+        Args:
+            loader (DataLoader): The data loader to sample from.
+            guide (AutoGuide): The guide to evaluate the proposals.
+            num_threads (int): Number of threads to use for sampling.
+            sub_batch_size (int, optional): Size of the sub-batch to use for sampling. If None, uses the batch size.
+            score_range (tuple): The range of scores to consider valid.
+        """
+        self.loader = loader
+        self.guide = guide
+        self.num_threads = num_threads
+        if sub_batch_size is None:
+            sub_batch_size = loader.batch_size
+        else:
+            assert sub_batch_size <= loader.batch_size, "sub_batch_size must be less than or equal to the loader's batch size."
+        self.sub_batch_size = sub_batch_size
+        self.score_range = score_range
+        if forward is None:
+            self.forward = standard_forward
+
+    @property
+    def dataset(self):
+        """ Get the dataset of the loader. """
+        return self.loader.dataset
+
+    @dataset.setter
+    def dataset(self, value):
+        """ Set the dataset of the loader. """
+        assert isinstance(value, dict), "Dataset must be a dictionary with 'inputs' and 'infos' keys."
+        assert 'inputs' in value and 'infos' in value, "Dataset must contain 'inputs' and 'infos' keys."
+        assert len(value['inputs']) == len(value['infos']), "Length of inputs must match length of infos."
+        self.loader.dataset = value
+
+    @property
+    def batch_size(self):
+        """ Get the batch size of the loader. """
+        return self.loader.batch_size
+
+    @batch_size.setter
+    def batch_size(self, value):
+        """ Set the batch size of the loader. """
+        self.loader.batch_size = value
+
+    @property
+    def n_epochs(self):
+        """ Get the number of epochs of the loader. """
+        return self.loader.n_epochs
+
+    def sample(self, agents, description_prefix=''):
+        """ Sample a batch of data from the loader and evaluate the agents.
+
+        Args:
+            agents (list): A list of trace.Modules (proposed parameters) to evaluate.
+
+        Returns:
+           batch (dict):
```

Comment:
The input data is included both here in the `batch` and in `samples:rollouts:x/info`, is that true?
If so, I suggest not return `batch`, and have another function to process `samples` to extract `batch`